apps/docusaurus/docs/guides/randomness.md

```diff
@@ -0,0 +1,314 @@
+malicious players can deploy their own contract to:
+
+1. call`decide_winner_internal()`;
+1. read the lottery result (assuming the `lottery` module has some getter functions for the result);
+1. abort if the result is not in their favor.
+   By repeatedly calling their own contract until a txn succeeds,
+   malicious users can bias the uniform distribution of the winner (dApp developer's initial design).
+   This is referred to as a _test-and-abort attack_.
+
+Aptos move compiler have been updated to prevent this attack for your contract safety:
+a randomness-dependent public function is treated as a compile error.
+If you have finished the steps in the ["build aptos-cli"](#build-aptos-cli-from-latest-source)) section,
+then your aptos-cli are equipped with the updated compiler.
+
+```rust
+module module_owner::lottery {
+    // Compile error!
+    public fun decide_winner_internal(lottery_state: &mut lottery_state) {
+        let n = std::vector::length(&lottery_state.players);
+        let winner_idx = aptos_framework::randomness::u64_range(0, n);
+        lottery_state.winner_idx = std::option::some(winner_idx);
+    }
+}
+```
+
+Not recommended, but if you intend to expose such a randomness-dependent function to the public, you can bypass the compiler check by annotating your function with `#[lint_allow_unsafe_randomness]`.
+
+```rust
+module module_owner::lottery {
+    // Can compile, but use it at your own risk!
+    #[lint_allow_unsafe_randomness]
+    public fun decide_winner_internal(lottery_state: &mut lottery_state) {
+        let n = std::vector::length(&lottery_state.players);
+        let winner_idx = aptos_framework::randomness::u64_range(0, n);
+        lottery_state.winner_idx = std::option::some(winner_idx);
+    }
+}
+```
+
+### Security warning: undergasing attacks, and how to prevent
+
+Imagine such a dApp. It defines a private entry function for a user to:
+
+1. toss a coin (gas cost: 9), then
+2. get a reward (gas cost: 10) if coin=1, or get multiple punishments (gas cost: 100) otherwise.
+
+A malicious user can control its account blanace so it covers at most 108 gas units (or set transaction parameter `max_gas=108`), and the punishing branch will never be reached.
+
+Formally, this is referred to as an _undergasing attack_,
+where an attacker can control how much gas is left for the entry function to execute,
+and so can arbitrarily decide to abort paths that cost more gas,
+biasing the outcome (i.e. effectively changing the distribution of random numbers).
+
+_WARNING: randomness API currently doesn’t prevent undergassing attacks._
+As a dApp developer, you need to be very careful in your design to avoid this type of attack.
+Here are some ideas of how to prevent undergasing attack generally.
+
+- Make your entry function gas independent from the randomness outcome.
+  Simplest example is to not “act” on the randomness outcome, i.e. read it and store it for later.
```

Comment:
You can link to the example in AIP-41 (my AIP-41 changes did not merge yet though)

Reply:
separate PR?